src/connectors/s3.rs

```diff
@@ -0,0 +1,414 @@
+
+    // Required later
+    async fn on_signal(
+        &mut self,
+        _signal: Event,
+        _ctx: &SinkContext,
+        _serializer: &mut EventSerializer,
+    ) -> Result<SinkReply> {
+        // Check if a new client is required or not.
+        if let Ok(new_client) = self.client_receiver.try_recv() {
+            self.client = Some(new_client);
+        }
+        Ok(SinkReply::default())
+    }
+
+    async fn on_stop(&mut self, ctx: &SinkContext) -> Result<()> {
+        // Commit the final upload.
+        self.complete_multipart(ctx).await?;
+        Ok(())
+    }
+
+    fn asynchronous(&self) -> bool {
+        true
+    }
+
+    fn auto_ack(&self) -> bool {
+        false
+    }
+}
+
+#[async_trait::async_trait]
+impl Connector for S3Connector {
+    /// Currently no source
+    async fn create_source(
+        &mut self,
+        _source_context: SourceContext,
+        _builder: SourceManagerBuilder,
+    ) -> Result<Option<SourceAddr>> {
+        Ok(None)
+    }
+
+    /// Stream the events to the bucket
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let (client_sender, client_receiver) = channel::bounded(2);
+
+        // set the sender and receiver for a new client
+        self.client_sender = Some(client_sender);
+
+        let s3_sink = S3Sink {
+            client_receiver,
+            bucket: self.config.bucket.clone(),
+            client: None,
+            buffer: Vec::with_capacity(self.config.min_part_size),
+            current_key: "".to_owned(),
+            parts: Vec::new(),
+            upload_id: "".to_owned(),
+            part_number: 0,
+            min_part_size: self.config.min_part_size,
+        };
+
+        let addr = builder.spawn(s3_sink, sink_context)?;
+        Ok(Some(addr))
+    }
+
+    async fn connect(&mut self, _ctx: &ConnectorContext, _attempt: &Attempt) -> Result<bool> {
+        // Create the client and establish the connection.
+        let s3_config = s3::config::Config::builder()
+            .credentials_provider(Credentials::new(
+                self.config.aws_access_key_id.clone(),
+                self.config.aws_secret_access_key.clone(),
+                None,
+                None,
+                "Environment",
+            ))
+            .region(Region::new(self.config.aws_region.clone()));
+
+        let s3_config = match &self.endpoint {
+            Some(uri) => s3_config.endpoint_resolver(Endpoint::immutable(uri.clone())),
+            None => (s3_config),
+        };
+
+        let client = S3Client::from_conf(s3_config.build());
+
+        // Check for the existence of the bucket.
+        let exist_bucket = client
+            .head_bucket()
+            .bucket(self.config.bucket.clone())
+            .send()
+            .await;
+
+        if exist_bucket.is_err() {
+            return Err(exist_bucket.unwrap_err().into());
+        }
```

Comment:
We don't want panicing code if we can avoid it lets rewrite this to:
```suggestion
        if let Err(e) = exist_bucket {
            return Err(exist_bucket.into());
        }
```